pkg/controller/releasemanager/plan/syncCanaryRules.go

```diff
@@ -0,0 +1,149 @@
+package plan
+
+import (
+	"context"
+	"fmt"
+
+	monitoringv1 "github.com/coreos/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/go-logr/logr"
+	picchuv1alpha1 "go.medium.engineering/picchu/pkg/apis/picchu/v1alpha1"
+	"go.medium.engineering/picchu/pkg/plan"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	CanaryAppLabel = "app"
+	CanaryTagLabel = "tag"
+	CanarySLOLabel = "slo"
+	CanaryLabel    = "canary"
+)
+
+type SyncCanaryRules struct {
+	App                         string
+	Namespace                   string
+	Tag                         string
+	Labels                      map[string]string
+	ServiceLevelObjectiveLabels picchuv1alpha1.ServiceLevelObjectiveLabels
+	ServiceLevelObjectives      []*picchuv1alpha1.ServiceLevelObjective
+}
+
+func (p *SyncCanaryRules) Apply(ctx context.Context, cli client.Client, scalingFactor float64, log logr.Logger) error {
+	prometheusRules, err := p.prometheusRules()
+	if err != nil {
+		return err
+	}
+
+	if len(prometheusRules.Items) > 0 {
+		for _, pr := range prometheusRules.Items {
+			if err := plan.CreateOrUpdate(ctx, log, cli, pr); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (p *SyncCanaryRules) prometheusRules() (*monitoringv1.PrometheusRuleList, error) {
+	prl := &monitoringv1.PrometheusRuleList{}
+	prs := []*monitoringv1.PrometheusRule{}
+
+	rule := p.prometheusRule()
+
+	for _, slo := range p.ServiceLevelObjectives {
+		if slo.ServiceLevelIndicator.Canary.Enabled {
+			config := SLOConfig{
+				SLO:    slo,
```

Comment:
Is this going to be the famous pointer in a for loop issue?

Reply:
I'm guessing no because canaryRules() doesn't return anything with the pointer. Still might be good to defend from future bugs by either copying or using the index.

Reply:
I switched these loops to all use the index.